### ISSUE TITLE
[Deferred Components] Deferred components download issue with mobile data while it works with wi-fi

### ISSUE THREAD

**Issue by enricoi** [flutter/flutter#117559]
Deferred Components works perfectly when using wi-fi. With mobile data when wi-fi is off almost every time components don't even start to download, or sometimes just one component of three start to download. Already tested on different devices and it isn't reliable. 

<details>
  <summary>Logs</summary>

[✓] Flutter (Channel stable, 3.3.9, on macOS 12.6 21G115 darwin-arm, locale it-IT)
    • Flutter version 3.3.9 on channel stable at /Users/---/development/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision b8f7f1f986 (4 weeks ago), 2022-11-23 06:43:51 +0900
    • Engine revision 8f2221fbef
    • Dart version 2.18.5
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/---/Library/Android/sdk
    • Platform android-33, build-tools 32.0.0
    • ANDROID_HOME = /Users/---/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode 2.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.73.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.56.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 12.6 21G115 darwin-arm
    • Chrome (web)    • chrome • web-javascript • Google Chrome 108.0.5359.124

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!

</details>

**Comment by darshankawar** [comment]
Thanks for the report @enricoi 
Can you provide a code sample and steps to replicate that shows the behavior along with logs for mobile data as compared to wifi ?

**Comment by enricoi** [comment]
Unfortunately I can only check the issue once it has been uploaded to the play store has internal test mode or in production, without possibility to receive error logs, or at least I don't know how to do it.
Steps to reproduce are: upload a build on play store with deferred components. Try it with mobile data and with wifi to notice the different behavior.

**Comment by darshankawar** [comment]
@enricoi 
Does it happen only while using deferred componenets or without it too ?
It could probably be related to ipv4 or ipv6 that you might need to confirm. What Android version are you seeing this behavior on ?

**Comment by enricoi** [comment]
The issue is that assets(images) of deferred components are not available because deferred component download doesn't start if in mobile data mode. Tested on Android 10, Android 11 and Android 13 with same behavior.

Plus on Android 13 if in wifi mode deferred components download start and complete well but to have assets available you have to restart the app.

**Comment by darshankawar** [comment]
Please check below issues and see if they help in your case:

https://github.com/flutter/flutter/issues/96046
https://github.com/flutter/flutter/issues/100855

**Comment by enricoi** [comment]
No they don't help me. I'm able to build with deferred components and upload It to play store without problem. Deferred components and relative assets download works perfect in wifi mode. The issue is on mobile data mode only.

**Comment by darshankawar** [comment]
> The issue is on mobile data mode only.

I am not completely sure how this would be related to flutter framework. Since it works on wifi but not with mobile data is probably related to the network speed. Can you try with a faster or more stable mobile data connection and see if it helps ?

**Comment by enricoi** [comment]
I've already tested with different mobile data provider and different devices with the same results. Speed test around 30 mb/s. I guess if it was related to network speed even the main package on play store should have problem on download but it completes in few seconds.

**Comment by darshankawar** [comment]
Thanks for the update @enricoi. Before we make this issue actionable, can you provide a log that shows the download details which shows the time delay using mobile network as compared to wifi ?

**Comment by enricoi** [comment]
Thanks @darshankawar. Please can you explain me how to do it with the internal test release uploaded on play store?

**Comment by darshankawar** [comment]
> Please can you explain me how to do it with the internal test release uploaded on play store?

Not from the playstore, but if you are locally able to see the behavior in `--release` mode, then I was looking for the logs between downloading via wifi as compared to mobile data.

**Comment by enricoi** [comment]
But `flutter run --release` generate an .app-release.apk and not the .aab file with deferred components.

**Comment by darshankawar** [comment]
You can run `flutter build appbundle` and pass below flag to use deferred components in it.

<img width="1307" alt="Screenshot 2022-12-29 at 6 16 50 PM" src="https://user-images.githubusercontent.com/67046386/209955425-af730d31-c374-42c7-9652-3af698e90cba.png">

**Comment by enricoi** [comment]
`flutter build appbundle` is what I use to build .aab to upload on play store. It works and by default use deferred components in it. How can I run the .aab in local?

**Comment by exaby73** [comment]
Hi @enricoi. You can check [this answer](https://stackoverflow.com/a/57628287/10845917) on Stackoverflow to deploy an aab file to Android

**Comment by enricoi** [comment]
Ok I've build and deploy with bundletool and I got this error:

`Failed to remove working directory with local testing splits. Your app might still have been installed correctly but have previous version of dynamic feature modules. If you see legacy versions of dynamic feature modules installed try to uninstall and install the app again.`

By the way if I open the installed app everything works right and deferred component assets are available.

**Comment by darshankawar** [comment]
> Steps to reproduce are: upload a build on play store with deferred components. Try it with mobile data and with wifi to notice the different behavior.

Coming back to this. Is it possible for you to give access of your project to us and then we can try to download using both (wifi and mobile data) to see the actual behavior ? You can add me (darshan@nevercode.io).

**Comment by enricoi** [comment]
I've send you an email last week. Have you received it?

**Comment by darshankawar** [comment]
Sorry, somehow I seem to have missed that email notification and there was also no update in this thread. I looked for the app in playstore but seem to see quite a few with same name. Which one among them should I try ?

<img width="356" alt="Screenshot 2023-01-10 at 11 26 53 AM" src="https://user-images.githubusercontent.com/67046386/211472705-77186902-4325-4a3a-8470-e6c8b919e944.png">

**Comment by enricoi** [comment]
All of them have the same issue. Choose whatever you prefer. Thanks

**Comment by darshankawar** [comment]
Ok, I installed the app using wifi and cellular data which worked without any issues.

<img width="396" alt="Screenshot 2023-01-11 at 11 29 58 AM" src="https://user-images.githubusercontent.com/67046386/211729503-268c3616-962e-4303-ba66-e424fc831d89.png">

**Comment by enricoi** [comment]
To check the issue you have to authenticate and then tap on world, "the
city" in example, and then tap on one story, "the station" in example. You
will see that if you are in wifi mode everything works correctly, other way
with mobile data you will see white screen due to image assets from
deferred components missing.

In alternative you can check the notification bar of your phone when you
are in authentication screen. If you are in wifi mode you will see three
deferred components download progress bar start and complete one after each
other. In mobile data mode that not happens at all almost everytime you try
or in very few case not all of the three components start download and
complete successfull.

Il Mer 11 Gen 2023, 07:02 darshankawar ***@***.***> ha
scritto:

**Comment by darshankawar** [comment]
Thanks for the update, but I am unable to properly verify this behavior, so going to keep this issue open for team's attention and would need additional triage.

P.S: @enricoi I tried to login using google account, but can't seem to find how to logout. Let me know so that I can logout :)

**Comment by enricoi** [comment]
@darshankawar To logout go in settings and then sign out.
Sorry, but why aren't you able to verify that behavior?

**Comment by enricoi** [comment]
@reidbaker I think that is more than a P5 new feature request. If deferred components don't work with mobile data and don't have a stable behavior they're pretty useless. It's a feature that already exist and doesn't work as it should do.

**Comment by Hixie** [comment]
That's fair. Unfortunately even with a higher priority, without being able to reproduce the problem it's really not clear how we can debug it.

Are you able to reproduce this with a minimal application that does nothing but try to show the problem?

**Comment by enricoi** [comment]
Hi @Hixie! 
unfortunately the issue is reproducible only after upload the bundle on play store.

By the way after further investigations I noticed that deferred components documentation at this [link](https://docs.flutter.dev/perf/deferred-components#step-1-dependencies-and-initial-project-setup) suggest to add this dependency `com.google.android.play:core:1.8.0`. As you could see [here](https://developer.android.com/reference/com/google/android/play/core/release-notes) is pretty obsolete (July 2020).

After that you could see many minor versions update with bug fix for Play Feature Delivery till 1.10.3 (January 2022), and one major version upgrade 2.0.0 (May 2022) with the Play Core Java and Kotlin Library that has been partitioned into multiple per-feature Android libraries including Play Feature Delivery Library.

So I tried to upgrade the dependency to see if the issue disappear. I was able to upgrade till play core 1.10.3 and components download seems to be better but still not stable. Tested with different devices and different android versions. It works always in wifi mode, it works sometimes with mobile data. Sometimes it start as soon as I call  `DeferredComponent.installDeferredComponent` and successfully complete, sometimes it start after a while (2 -3 minutes) and complete, sometimes it start the download of just one or two components out of three, sometimes doesn't start at all.

When I tried to add the most recent dependency 'com.google.android.play:app-update:2.0.1' the app crash with this error
in debug mode:
```
E/AndroidRuntime(28454): FATAL EXCEPTION: main
E/AndroidRuntime(28454): Process: com.languakids.english_on_the_farm, PID: 28454
E/AndroidRuntime(28454): java.lang.NoSuchMethodError: No interface method startInstall(Lcom/google/android/play/core/splitinstall/SplitInstallRequest;)Lcom/google/android/play/core/tasks/Task; in class Lcom/google/android/play/core/splitinstall/SplitInstallManager; or its super classes (declaration of 'com.google.android.play.core.splitinstall.SplitInstallManager' appears in /data/app/~~V9HRkYRrdW8nreUZkV5zUA==/com.languakids.english_on_the_farm-R5otvxtVnn9xHN92iZBzag==/base.apk!classes6.dex)
E/AndroidRuntime(28454): 	at io.flutter.embedding.engine.deferredcomponents.PlayStoreDeferredComponentManager.installDeferredComponent(PlayStoreDeferredComponentManager.java:314)
E/AndroidRuntime(28454): 	at io.flutter.embedding.engine.systemchannels.DeferredComponentChannel$1.onMethodCall(DeferredComponentChannel.java:54)
E/AndroidRuntime(28454): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:262)
E/AndroidRuntime(28454): 	at io.flutter.embedding.engine.dart.DartMessenger.invokeHandler(DartMessenger.java:295)
E/AndroidRuntime(28454): 	at io.flutter.embedding.engine.dart.DartMessenger.lambda$dispatchMessageToQueue$0$DartMessenger(DartMessenger.java:319)
E/AndroidRuntime(28454): 	at io.flutter.embedding.engine.dart.DartMessenger$$ExternalSyntheticLambda0.run(Unknown Source:12)
E/AndroidRuntime(28454): 	at android.os.Handler.handleCallback(Handler.java:942)
E/AndroidRuntime(28454): 	at android.os.Handler.dispatchMessage(Handler.java:99)
E/AndroidRuntime(28454): 	at android.os.Looper.loopOnce(Looper.java:226)
E/AndroidRuntime(28454): 	at android.os.Looper.loop(Looper.java:313)
E/AndroidRuntime(28454): 	at android.app.ActivityThread.main(ActivityThread.java:8741)
E/AndroidRuntime(28454): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(28454): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:571)
E/AndroidRuntime(28454): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1067)
I/ProviderInstaller(28454): Installed default security provider GmsCore_OpenSSL
Lost connection to device.
```

I've also tried to build and upload this version on play store and it crash as soon as I call `DeferredComponent.installDeferredComponent`

I don't know if this could be helpful to solve the problem. Maybe with the most recent dependency deferred components could works as expected.

**Comment by enricoi** [comment]
With previous version of play core dependencies(1.8.0) also received this kind of crash from google play store dashboard. Don't know if it's related but it concern deferred components.

```
Exception java.lang.ClassCastException:
  at io.flutter.embedding.engine.deferredcomponents.PlayStoreDeferredComponentManager.lambda$installDeferredComponent$1$PlayStoreDeferredComponentManager (PlayStoreDeferredComponentManager.java:329)
  at io.flutter.embedding.engine.deferredcomponents.PlayStoreDeferredComponentManager$$ExternalSyntheticLambda0.onFailure
  at com.google.android.play.core.tasks.c.run
  at android.os.Handler.handleCallback (Handler.java:938)
  at android.os.Handler.dispatchMessage (Handler.java:99)
  at android.os.Looper.loop (Looper.java:246)
  at android.app.ActivityThread.main (ActivityThread.java:8653)
  at java.lang.reflect.Method.invoke
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:602)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1130)
```

**Comment by Lokhandwalah** [comment]
> Hi @Hixie! unfortunately the issue is reproducible only after upload the bundle on play store.
> 
> By the way after further investigations I noticed that deferred components documentation at this [link](https://docs.flutter.dev/perf/deferred-components#step-1-dependencies-and-initial-project-setup) suggest to add this dependency `com.google.android.play:core:1.8.0`. As you could see [here](https://developer.android.com/reference/com/google/android/play/core/release-notes) is pretty obsolete (July 2020).
> 
> After that you could see many minor versions update with bug fix for Play Feature Delivery till 1.10.3 (January 2022), and one major version upgrade 2.0.0 (May 2022) with the Play Core Java and Kotlin Library that has been partitioned into multiple per-feature Android libraries including Play Feature Delivery Library.
> 
> So I tried to upgrade the dependency to see if the issue disappear. I was able to upgrade till play core 1.10.3 and components download seems to be better but still not stable. Tested with different devices and different android versions. It works always in wifi mode, it works sometimes with mobile data. Sometimes it start as soon as I call `DeferredComponent.installDeferredComponent` and successfully complete, sometimes it start after a while (2 -3 minutes) and complete, sometimes it start the download of just one or two components out of three, sometimes doesn't start at all.
> 
> When I tried to add the most recent dependency 'com.google.android.play:feature-delivery:2.0.1' the app crash with this error in debug mode:
> 
> ```
> E/AndroidRuntime(28454): FATAL EXCEPTION: main
> E/AndroidRuntime(28454): Process: com.languakids.english_on_the_farm, PID: 28454
> E/AndroidRuntime(28454): java.lang.NoSuchMethodError: No interface method startInstall(Lcom/google/android/play/core/splitinstall/SplitInstallRequest;)Lcom/google/android/play/core/tasks/Task; in class Lcom/google/android/play/core/splitinstall/SplitInstallManager; or its super classes (declaration of 'com.google.android.play.core.splitinstall.SplitInstallManager' appears in /data/app/~~V9HRkYRrdW8nreUZkV5zUA==/com.languakids.english_on_the_farm-R5otvxtVnn9xHN92iZBzag==/base.apk!classes6.dex)
> E/AndroidRuntime(28454): 	at io.flutter.embedding.engine.deferredcomponents.PlayStoreDeferredComponentManager.installDeferredComponent(PlayStoreDeferredComponentManager.java:314)
> E/AndroidRuntime(28454): 	at io.flutter.embedding.engine.systemchannels.DeferredComponentChannel$1.onMethodCall(DeferredComponentChannel.java:54)
> E/AndroidRuntime(28454): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:262)
> E/AndroidRuntime(28454): 	at io.flutter.embedding.engine.dart.DartMessenger.invokeHandler(DartMessenger.java:295)
> E/AndroidRuntime(28454): 	at io.flutter.embedding.engine.dart.DartMessenger.lambda$dispatchMessageToQueue$0$DartMessenger(DartMessenger.java:319)
> E/AndroidRuntime(28454): 	at io.flutter.embedding.engine.dart.DartMessenger$$ExternalSyntheticLambda0.run(Unknown Source:12)
> E/AndroidRuntime(28454): 	at android.os.Handler.handleCallback(Handler.java:942)
> E/AndroidRuntime(28454): 	at android.os.Handler.dispatchMessage(Handler.java:99)
> E/AndroidRuntime(28454): 	at android.os.Looper.loopOnce(Looper.java:226)
> E/AndroidRuntime(28454): 	at android.os.Looper.loop(Looper.java:313)
> E/AndroidRuntime(28454): 	at android.app.ActivityThread.main(ActivityThread.java:8741)
> E/AndroidRuntime(28454): 	at java.lang.reflect.Method.invoke(Native Method)
> E/AndroidRuntime(28454): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:571)
> E/AndroidRuntime(28454): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1067)
> I/ProviderInstaller(28454): Installed default security provider GmsCore_OpenSSL
> Lost connection to device.
> ```
> 
> I've also tried to build and upload this version on play store and it crash as soon as I call `DeferredComponent.installDeferredComponent`
> 
> I don't know if this could be helpful to solve the problem. Maybe with the most recent dependency deferred components could works as expected.

I am facing this exact issue in my project as well. Did you find any workaround?

**Comment by enricoi** [comment]
Hi @Lokhandwalah,
No I didn't. Unfortunately deferred components isn't reliable. until it will I'm using a different approach with flutter_play_asset_delivery. To be honest I don't test deferred components since january 2023.

**Comment by KonstantinKorotaev** [comment]
Any news on this issue?
